### PR TITLE
add lock action and relative date preprocess

### PR DESCRIPTION
**Pull request by connorjclark**
example usage:

```yaml
filters:
  - type: issue
    criteria:
      state: closed
      locked: false
      updatedAt:
        $lt: '%%date(3 months ago)%%'
actions:
  - type: add_comment
    body: >
      This issue has been resolved and there has been no further activity, so we're gonna lock it.
      Please open a new issue if you have a similar issue, and be sure to fill out the template
      and reference this issue.
  - type: lock

```

### CODE REVIEW

File: lib/filter.js [hunk 1]
```diff
@@ -70,6 +71,27 @@ class Filter {
     return value => conditions.every(condition => condition(value))
   }
 
+  static preprocess(object, issue) {
+    if (typeof object === 'undefined') {
+      object = {text: ''}
```
Comment (connorjclark) [inline lib/filter.js]: idk what this does, maybe should just be in the latest activity filter?

File: lib/filter.js [hunk 2]
```diff
@@ -70,6 +71,27 @@ class Filter {
...
+    }
+
+    const replacements = {
+      author: issue.user && issue.user.login,
+      date: relativeTimePhrase => chrono.parseDate(relativeTimePhrase),
+    }
+
+    _.forEach(replacements, (value, name) => {
+      if (typeof value === 'function') {
+        Filter.replacePropertyDynamic(object, new RegExp(`%%${name}\\((.*)\\)%%`), value)
```
Comment (patrickhulce) [inline lib/filter.js]:
```suggestion
        Filter.replacePropertyDynamic(object, new RegExp(`%%${name}\\((.*?)\\)%%`), value)
```

File: lib/action.js [hunk 3]
```diff
@@ -32,6 +33,8 @@ class Action {
         return github.removeLabel(issue, this._action.label)
       case Action.TYPES.CLOSE:
         return github.closeIssue(issue)
+      case Action.TYPES.LOCK:
+        return github.lockIssue(issue, this._action.label)
```
Comment (patrickhulce) [inline lib/action.js]: should this be `reason` instead? would be great to throw this into the example too and/or link to https://developer.github.com/v3/issues/#parameters-4 for the options :)

File: lib/filter.js (hunk 1 above)
Comment (patrickhulce) [reply lib/filter.js]: it's just defaulting to a filter that will always pass, it's fine to leave here though maybe add a comment to this effect :)

File: lib/github.js [hunk 4]
```diff
@@ -157,12 +157,23 @@ class GitHubApi {
...
+  }
+
   static transformIssue(issue) {
     return {
       text: [issue.title, issue.body].join('\n'),
       title: issue.title,
       body: issue.body,
       state: issue.state,
+      locked: issue.locked,
+      updatedAt: new Date(issue.updated_at),
```
Comment (patrickhulce) [inline lib/github.js]: nit: the `createdAt` `updatedAt` props always feel like they belong at the end of object definitions :)

File: lib/github.js [hunk 5]
```diff
@@ -157,12 +157,23 @@ class GitHubApi {
     )
   }
 
+  lockIssue(issue, lockReason) {
+    return this.dryRunGuardedRequest(
+      `/issues/${issue.number}/lock`,
+      'put',
+      // eslint-disable-next-line camelcase
+      request => request.type('json').send({lockReason: lockReason || 'resolved'})
```
Comment (patrickhulce) [inline lib/github.js]: is this supposed to be `lock_reason`?
```suggestion
      request => request.type('json').send({lock_reason: lockReason || 'resolved'})
```

**Pull request merged**